google-cloud-datastore/lib/google/cloud/datastore/filter.rb

```diff
@@ -0,0 +1,208 @@
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   filter = Google::Cloud::Filter.new("done", "=", false)
+        #                                 .and("priority", ">=", 4)
+        #
+        # @example Join the filter with a filter object
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   filter_1 = Google::Cloud::Filter.new("done", "=", false)
+        #   filter_2 = Google::Cloud::Filter.new("priority", ">=", 4)
+        #
+        #   filter = filter_1.and(filter_2)
+        #
+        def and name_or_filter, operator = nil, value = nil
+          combine_filters composite_filter_and, name_or_filter, operator, value
+        end
+
+        ##
+        # Joins two filters with an OR operator.
+        #
+        # @overload or(name, operator, value)
+        #   Joins the filter with a property filter
+        #   @param name [String]
+        #   @param operator [String]
+        #   @param value
+        #
+        # @overload or(filter)
+        #   Joins the filter with a Filter object
+        #   @param flter [Filter]
+        #
+        # @example Join the filter with a property filter
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   filter = Google::Cloud::Filter.new("done", "=", false)
+        #                                 .or("priority", ">=", 4)
+        #
+        # @example Join the filter with a filter object
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   filter_1 = Google::Cloud::Filter.new("done", "=", false)
+        #   filter_2 = Google::Cloud::Filter.new("priority", ">=", 4)
+        #
+        #   filter = filter_1.or(filter_2)
+        #
+        def or name_or_filter, operator = nil, value = nil
+          combine_filters composite_filter_or, name_or_filter, operator, value
+        end
+
+        # @private
+        def to_grpc
+          @grpc
+        end
+
+        protected
+
+        ##
+        # @private
+        def combine_filters composite_filter, name_or_filter, operator, value
+          composite_filter.composite_filter.filters << to_grpc
+          composite_filter.composite_filter.filters << if name_or_filter.is_a? Google::Cloud::Datastore::Filter
+                                                         name_or_filter.to_grpc
+                                                       else
+                                                         create_property_filter name_or_filter, operator, value
+                                                       end
+          self.class.new("", "", "").tap do |f|
+            f.grpc = composite_filter
+          end
+        end
+
+        ##
+        # @private
+        def composite_filter_and
```

Comment:
Just a thought
How about we combine both `composite_filter_and` and `composite_filter_or` with say `composite_filter` and let it accept an argument which will be the operation `:OR` or `:AND` ?

Reply:
I considered this, but I wanted to avoid the `if...else` nesting & keep the code linear. So I choose 2 helper methods instead.

Reply:
No, you won't have to use `if` at all.

Method definition: 

```
def composite_filter op: :AND
  Google::Cloud::Datastore::V1::Filter.new(
            composite_filter: Google::Cloud::Datastore::V1::CompositeFilter.new(op: op)
          )
end
```

Invocation:

```
composite_filter :OR
```

Reply:
Based on the offline discussion, I'm keeping this code as it is (reason: Rubocop doesn't allow method names to be shorter than 3 characters, and `op` is only 2 characters long)